Euclidean.lua

```diff
@@ -32,36 +33,146 @@ function Euclidean:reset(stdv)
 end
 
 function Euclidean:updateOutput(input)
-   self.output:zero()
-   for o = 1,self.weight:size(2) do
-      self.output[o] = input:dist(self.weight:select(2,o))
+   -- lazy initialize buffers
+   self._input = self._input or input.new()
+   self._weight = self._weight or self.weight.new()
+   self._expand = self._expand or self.output.new()
+   self._expand2 = self._expand2 or self.output.new()
+   self._repeat = self._repeat or self.output.new()
+   self._repeat2 = self._repeat2 or self.output.new()
+   
+   local inputSize, outputSize = self.weight:size(1), self.weight:size(2)
+   
+   -- y_j = || w_j - x || = || x - w_j ||
+   if input:dim() == 1 then
+      self._input:reshape(input, inputSize, 1)
```

Comment:
reshape means a copy for sure. can we make a :viewAs out of this line by any chance?